src/rpc/bobtail.cpp

```diff
@@ -0,0 +1,245 @@
+#include "core_io.h"
+#include "dstencode.h"
+#include "init.h"
+#include "main.h"
+#include "net.h"
+#include "parallel.h"
+#include "pow.h"
+#include "rpc/server.h"
+#include "txadmission.h"
+#include "txmempool.h"
+#include "ui_interface.h"
+#include "util.h"
+#include "utilstrencodings.h"
+#include "validationinterface.h"
+
+#include <cstdlib>
+#include <stdint.h>
+
+#include <boost/assign/list_of.hpp>
+#include <boost/shared_ptr.hpp>
+
+extern CBobtailDagSet bobtailDagSet;
+
+UniValue generateBobtailBlocks(boost::shared_ptr<CReserveScript> coinbaseScript,
+    int nGenerate,
+    uint64_t nMaxTries,
+    bool keepScript,
+    int weak_mode)
+{
+    static const int nInnerLoopCount = 0x10000;
+
+    unsigned int nExtraNonce = 0;
+    UniValue blockHashes(UniValue::VARR);
+
+    int numblocks = 0;
+
+    while (numblocks < nGenerate)
+    {
+        std::unique_ptr<CSubBlockTemplate> pblocktemplate;
+        {
+            TxAdmissionPause lock; // flush any tx waiting to enter the mempool
+            pblocktemplate = SubBlockAssembler(Params()).CreateNewSubBlock(coinbaseScript->reserveScript);
+        }
+        if (!pblocktemplate.get())
+            throw JSONRPCError(RPC_INTERNAL_ERROR, "Couldn't create new block");
+        LOG(WB, "Using delta block for RPC generate.\n");
+        CSubBlock *pblock = pblocktemplate->subblock.get();
+        {
+            // LOCK(cs_main);
+            IncrementExtraNonce(pblock, nExtraNonce);
+        }
+
+        // Generally look for weak PoW
+        while (nMaxTries > 0 && pblock->nNonce < nInnerLoopCount &&
+               !CheckProofOfWork(pblock->GetHash(), weakPOWfromPOW(pblock->nBits), Params().GetConsensus(), true))
+        {
+            ++pblock->nNonce;
+            --nMaxTries;
+        }
+        if (nMaxTries == 0)
+            break;
+        if (pblock->nNonce == nInnerLoopCount)
+            continue;
+
+        // Now check if Bobtail PoW is also satisfied
+        if (CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus()))
+        {
+            // In we are mining our own block or not running in parallel for any reason
+            // we must terminate any block validation threads that are currently running,
+            // Unless they have more work than our own block or are processing a chain
+            // that has more work than our block.
+            PV->StopAllValidationThreads(pblock->GetBlockHeader().nBits);
+
+            if (!ProcessNewSubBlock(*pblock))
+            {
+                throw JSONRPCError(RPC_INTERNAL_ERROR, "ProcessNewSubBlock, subblock not accepted");
+            }
+
+            // mark script as important because it was used at least for one coinbase output if the script came from the
+            // wallet
+            if (keepScript)
+            {
+                coinbaseScript->KeepScript();
+            }
+            numblocks++;
+        }
+        blockHashes.push_back(pblock->GetHash().GetHex());
```

Comment:
Same here, I think blockHashes should be updated outside the conditional. That said, we should be adding bobtailblocks inside this conditional to something like bobtailBlockHashes. It's fine to just add a TODO for that at this point.

Reply:
see comment above. i dont think there should be a change here anymore